File: frontend/src/services/authService.js
import client from "../api/client";

export async function login(email, password) {
  const { data } = await client.post("/login", { email, password });
  return data.data; // { access_token, token_type }
}

export async function fetchProfile() {
  const { data } = await client.get("/profile");
  return data.data; // { id, full_name, email, is_admin, is_active, created_at }
}

export async function changePassword({ current_password, new_password, confirm_password }) {
  const { data } = await client.post("/change-password", {
    current_password,
    new_password,
    confirm_password,
  });
  return data.message;
}

File: backend/app/services/auth_service.py
from sqlalchemy.orm import Session

from app.core.security import verify_password, hash_password, create_access_token
from app.models.teacher import Teacher
from app.repositories.teacher_repository import TeacherRepository
from app.schemas.auth import LoginRequest, TokenResponse, ChangePasswordRequest
from app.services.exceptions import AuthenticationError, ValidationError


class AuthService:
    def __init__(self, db: Session):
        self.teacher_repo = TeacherRepository(db)

    def login(self, payload: LoginRequest) -> TokenResponse:
        teacher = self.teacher_repo.get_by_email(payload.email)
        if not teacher or not verify_password(payload.password, teacher.password_hash):
            raise AuthenticationError("Invalid email or password.")
        if not teacher.is_active:
            raise AuthenticationError("This account has been deactivated.")

        token = create_access_token(
            subject=str(teacher.id),
            extra_claims={"email": teacher.email, "is_admin": teacher.is_admin},
        )
        return TokenResponse(access_token=token)

    def change_password(self, teacher: Teacher, payload: ChangePasswordRequest) -> None:
        if not verify_password(payload.current_password, teacher.password_hash):
            raise AuthenticationError("Current password is incorrect.")
        if payload.new_password != payload.confirm_password:
            raise ValidationError("New password and confirmation don't match.")

        teacher.password_hash = hash_password(payload.new_password)
        self.teacher_repo.commit_refresh(teacher)
